# Pandas GroupBy Operations (Lab 1)

**Learning Objectives:**
  * Gain an introduction to the `GroupBy` operation of the *pandas* library
  * Slice, dice and summarize data within a `DataFrame`
  

## Library import

The following line imports the *pandas* library

In [22]:
import pandas as pd


## Data loading and DataFrame creation


. The following example loads a file with the famous titanic data. Run the following cell to load the data and create your first `DataFrame`

In [23]:
titanic = pd.read_csv("https://raw.githubusercontent.com/thousandoaks/Python4DS101/master/data/titanic.csv", sep=",")


Let's display the first few records of a `DataFrame`:

In [24]:
titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

The data consists of the following data columns:

* PassengerId: Id of every passenger.

* Survived: This feature have value 0 and 1. 0 for not survived and 1 for survived.

* Pclass: There are 3 classes: Class 1, Class 2 and Class 3.

* Name: Name of passenger.

* Sex: Gender of passenger.

* Age: Age of passenger.

* SibSp: Indication that passenger have siblings and spouse.

* Parch: Whether a passenger is alone or have family.

* Ticket: Ticket number of passenger.

* Fare: Indicating the fare.

* Cabin: The cabin of passenger.

* Embarked: The embarked category.





## GroupBy Operations

GroupBy Operations follow the so-called Split-Apply-Combine paradigm whereby a `DataFrame` is: (1) split into groups, (2) a function is applied to each group and (3) the result are combined into a new `DataFrame` object.



<img src='https://pandas.pydata.org/docs/_images/06_groupby.svg'>

## We can groupby over several columns to summarize valuable information

In [25]:
# We split the dataset based on two columns: Sex and Survived. Then we apply the count() function to each group. Finally the results are combined into a new DataFrame
titanic.groupby(['Sex','Survived']).count()

PassengerId  Pclass  Name  Age  SibSp  Parch  Ticket  Fare  \
Sex    Survived                                                               
female 0                  81      81    81   64     81     81      81    81   
       1                 233     233   233  197    233    233     233   233   
male   0                 468     468   468  360    468    468     468   468   
       1                 109     109   109   93    109    109     109   109   

                 Cabin  Embarked  
Sex    Survived                   
female 0             6        81  
       1            91       231  
male   0            62       468  
       1            45       109

#### From the previous result we observe that most of the females survived (233). On the other hand most of the males did not survive (468).

In [26]:
titanic.groupby(['Sex','Pclass']).count()

PassengerId  Survived  Name  Age  SibSp  Parch  Ticket  Fare  \
Sex    Pclass                                                                 
female 1                94        94    94   85     94     94      94    94   
       2                76        76    76   74     76     76      76    76   
       3               144       144   144  102    144    144     144   144   
male   1               122       122   122  101    122    122     122   122   
       2               108       108   108   99    108    108     108   108   
       3               347       347   347  253    347    347     347   347   

               Cabin  Embarked  
Sex    Pclass                   
female 1          81        92  
       2          10        76  
       3           6       144  
male   1          95       122  
       2           6       108  
       3           6       347

#### From the previous result we observe that the majority of passengers, both males and females, travelled in third class

In [27]:
titanic.groupby(['Pclass','Survived']).count()

PassengerId  Name  Sex  Age  SibSp  Parch  Ticket  Fare  \
Pclass Survived                                                            
1      0                  80    80   80   64     80     80      80    80   
       1                 136   136  136  122    136    136     136   136   
2      0                  97    97   97   90     97     97      97    97   
       1                  87    87   87   83     87     87      87    87   
3      0                 372   372  372  270    372    372     372   372   
       1                 119   119  119   85    119    119     119   119   

                 Cabin  Embarked  
Pclass Survived                   
1      0            59        80  
       1           117       134  
2      0             3        97  
       1            13        87  
3      0             6       372  
       1             6       119

#### From the previous result we observe that the majority of passengers travelling in first class survived whereas the majority of passengers travelling on third class perished

In [28]:
# We split the dataset according to Pclass and Survived, then compute the mean of the column "Age" for each group.
titanic.groupby(['Pclass','Survived'])['Age'].mean()

Pclass  Survived
1       0           43.695312
        1           35.368197
2       0           33.544444
        1           25.901566
3       0           26.555556
        1           20.646118
Name: Age, dtype: float64

In [30]:
# We split the dataset according to Pclass and Survived, then compute the minimum value of the column "Fare" for each group.
titanic.groupby(['Pclass','Survived'])['Fare'].min()

Pclass  Survived
1       0            0.0000
        1           25.9292
2       0            0.0000
        1           10.5000
3       0            0.0000
        1            0.0000
Name: Fare, dtype: float64

In [31]:
# We split the dataset according to Pclass and Survived, then compute the maximum value of the column "Fare" for each group.
titanic.groupby(['Pclass','Survived'])['Fare'].max()

Pclass  Survived
1       0           263.0000
        1           512.3292
2       0            73.5000
        1            65.0000
3       0            69.5500
        1            56.4958
Name: Fare, dtype: float64